lib/commands/touch.js

```diff
@@ -0,0 +1,274 @@
+  if (ref) {
+    let opts = ref.options || {};
+    if (opts.element) {
+      // we retrieve the element location, might be useful in
+      // case the element becomes invalid
+      let loc = await this.getLocationInView(opts.element);
+      let size = await this.getSize(opts.element);
+      release.options = {
+        element: opts.element,
+        x: loc.x + size.width / 2,
+        y: loc.y + size.height / 2
+      };
+    }
+    if (opts.x && opts.y) {
+      release.options = _.pick(opts, 'x', 'y');
+    }
+  }
+  return release;
+};
+
+// Perform one gesture
+helpers.performGesture = async function (gesture) {
+  let res = await this.doTouchAction(gesture.action, gesture.options || {});
+  // sometime the element is not available when releasing, retry without it
+  if (res && res.status === 7 && gesture.action === 'release' && gesture.options.element) {
+    delete gesture.options.element;
+    log.debug(`retrying release without element opts: ${gesture.options}.`);
+    return await this.doTouchAction(gesture.action, gesture.options || {});
+  }
+  // otherwise continue normally
+  return res;
+};
+
+commands.performTouch = async function (gestures) {
+  // press-wait-moveTo-release is `swipe`, so use native method
+  if (gestures.length === 4 &&
+      gestures[0].action === 'press' &&
+      gestures[1].action === 'wait' &&
+      gestures[2].action === 'moveTo' &&
+      gestures[3].action === 'release') {
+
+      let swipeOpts = await this.getSwipeOptions(gestures);
+      return await this.swipe(swipeOpts.startX, swipeOpts.startY, swipeOpts.endX,
+                              swipeOpts.endY, swipeOpts.duration, swipeOpts.touchCount,
+                              swipeOpts.element);
+  }
+  let actions = _.pluck(gestures, "action");
+
+  if (actions[0] === 'longPress' && actions[1] === 'moveTo' && actions[2] === 'release') {
+    // some things are special
+    return await this.doTouchDrag(gestures);
+  } else {
+    // `press` without a wait is too slow and gets interpretted as a `longPress`
+    if (actions[actions.length - 2] === 'press' && actions[actions.length - 1] === 'release') {
+      actions[actions.length - 2] = 'tap';
+      gestures[gestures.length - 2].action = 'tap';
+    }
+
+    // the `longPress` and `tap` methods release on their own
+    if ((actions[actions.length - 2] === 'tap' ||
+      actions[actions.length - 2] === 'longPress') && actions[actions.length - 1] === 'release') {
+      gestures.pop();
+      actions.pop();
+    }
+
+    // fix release action then perform all actions
+    if (actions[actions.length - 1] === 'release') {
+      actions[actions.length - 1] = await this.fixRelease(gestures, actions);
+    }
+
+    let fixedGestures = await this.parseTouch(gestures, false);
+    for (let g of fixedGestures) {
+      await this.performGesture(g);
+    }
+  }
+};
+
+commands.parseTouch = async function (gestures, multi) {
+  if (multi && _.last(gestures).action === 'release') {
```

Comment:
comment about why we discard the last gesture if it's a release. is it because the automation backend does this for us automatically, or what?